salamander4/trace-cmd/LTS/trace-cmd-libtracecmd-1.5.1/Documentation/trace-cmd/trace-cmd.1.txt
TRACE-CMD(1)
============

NAME
----
trace-cmd - interacts with Ftrace Linux kernel internal tracer

SYNOPSIS
--------
*trace-cmd* 'COMMAND' ['OPTIONS']

DESCRIPTION
-----------
The trace-cmd(1) command interacts with the Ftrace tracer that is built inside
the Linux kernel. It interfaces with the Ftrace specific files found in the
debugfs file system under the tracing directory. A 'COMMAND' must be
specified to tell trace-cmd what to do.


COMMANDS
--------

  record  - record a live trace and write a trace.dat file to the
            local disk or to the network.

  set     - set a ftrace configuration parameter.

  report  - reads a trace.dat file and converts the binary data to a
            ASCII text readable format.

  stream  - Start tracing and read the output directly

  profile - Start profiling and read the output directly

  hist    - show a histogram of the events.

  stat    - show tracing (ftrace) status of the running system

  options - list the plugin options that are available to *report*

  start   - start the tracing without recording to a trace.dat file.

  stop    - stop tracing (only disables recording, overhead of tracer
            is still in effect)

  restart - restart tracing from a previous stop (only effects recording)

  extract - extract the data from the kernel buffer and create a trace.dat
            file.

  show    - display the contents of one of the Ftrace Linux kernel tracing files

  reset   - disables all tracing and gives back the system performance.
            (clears all data from the kernel buffers)

  clear   - clear the content of the Ftrace ring buffers.

  split   - splits a trace.dat file into smaller files.

  list    - list the available plugins or events that can be recorded.

  listen  - open up a port to listen for remote tracing connections.

  agent   - listen on a vsocket for trace clients

  setup-guest - create FIFOs for tracing guest VMs

  restore - restore the data files of a crashed run of trace-cmd record

  snapshot- take snapshot of running trace

  stack   - run and display the stack tracer

  check-events - parse format strings for all trace events and return
                 whether all formats are parseable

  convert   - convert trace files

  attach   - attach a host trace.dat file to a guest trace.dat file

  dump    - read out the meta data from a trace file

OPTIONS
-------

*-h*, --help::
    Display the help text.

Other options see the man page for the corresponding command.

SEE ALSO
--------
trace-cmd-record(1), trace-cmd-report(1), trace-cmd-hist(1), trace-cmd-start(1),
trace-cmd-stop(1), trace-cmd-extract(1), trace-cmd-reset(1),
trace-cmd-restore(1), trace-cmd-stack(1), trace-cmd-convert(1),
trace-cmd-split(1), trace-cmd-list(1), trace-cmd-listen(1),
trace-cmd.dat(5), trace-cmd-check-events(1), trace-cmd-stat(1),
trace-cmd-attach(1)

AUTHOR
------
Written by Steven Rostedt, <rostedt@goodmis.org>

RESOURCES
---------
https://git.kernel.org/pub/scm/utils/trace-cmd/trace-cmd.git/

COPYING
-------
Copyright \(C) 2010 Red Hat, Inc. Free use of this software is granted under
the terms of the GNU Public License (GPL).



salamander4/trace-cmd/LTS/trace-cmd-v3.2/tracecmd/trace-cmd.c
// SPDX-License-Identifier: GPL-2.0
/*
 * Copyright (C) 2008, 2009, 2010 Red Hat Inc, Steven Rostedt <srostedt@redhat.com>
 *
 */
#include <stdio.h>
#include <string.h>
#include <getopt.h>
#include <unistd.h>
#include <dirent.h>
#include <errno.h>
#include <stdlib.h>
#include <sys/syscall.h>

#include "trace-local.h"

int silence_warnings;
int show_status;

#ifndef gettid
#define gettid() syscall(__NR_gettid)
#endif

void warning(const char *fmt, ...)
{
	va_list ap;

	if (silence_warnings)
		return;

	if (errno)
		perror("trace-cmd");
	errno = 0;

	va_start(ap, fmt);
	fprintf(stderr, "  ");
	vfprintf(stderr, fmt, ap);
	va_end(ap);

	fprintf(stderr, "\n");
}

void *malloc_or_die(unsigned int size)
{
	void *data;

	data = malloc(size);
	if (!data)
		die("malloc");
	return data;
}

/* Same as strtok_r(), but allows empty tokens */
char *strparse(char *str, char delim, char **save)
{
	char *next;

	if (!str) {
		str = *save;
		if ((*save)[0] == '\0')
			return NULL;
	}

	next = strchr(str, delim);
	if (next) {
		*next = '\0';
		*save = next + 1;
	} else {
		*save = str + strlen(str);
	}
	return str;
}

void tracecmd_debug(const char *fmt, ...)
{
	va_list ap;

	if (!tracecmd_get_debug())
		return;

	va_start(ap, fmt);
	printf("[%d] ", (int)gettid());
	vprintf(fmt, ap);
	va_end(ap);
}

static struct trace_log_severity {
	int		id;
	const char	*name;
} log_severity[] = {
	{ .id = TEP_LOG_NONE, .name = "none" },
	{ .id = TEP_LOG_CRITICAL, .name = "crit" },
	{ .id = TEP_LOG_ERROR, .name = "err" },
	{ .id = TEP_LOG_WARNING, .name = "warn" },
	{ .id = TEP_LOG_INFO, .name = "info" },
	{ .id = TEP_LOG_DEBUG, .name = "debug" },
	{ .id = TEP_LOG_ALL, .name = "all" },
};

void trace_set_loglevel(int level)
{
	tracecmd_set_loglevel(level);
	tracefs_set_loglevel(level);
	tep_set_loglevel(level);
}

int trace_set_verbose(char *level)
{
	int id;

	/* Default level is info */
	if (!level)
		level = "info";

	if (isdigit(level[0])) {
		id = atoi(level);
		if (id >= TEP_LOG_NONE) {
			if (id > TEP_LOG_ALL)
				id = TEP_LOG_ALL;
			trace_set_loglevel(id);
			return 0;
		}
	} else {
		int size = ARRAY_SIZE(log_severity);
		int i;

		for (i = 0; i < size; i++) {
			if (!strncmp(level, log_severity[i].name, strlen(log_severity[i].name))) {
				trace_set_loglevel(log_severity[i].id);
				return 0;
			}
		}
	}

	return -1;
}

/**
 * struct command
 * @name command name
 * @run function to execute on command `name`
 */
struct command {
	char *name;
	void (*run)(int argc, char **argv);
};


/**
 * Lookup table that maps command names to functions
 */
struct command commands[] = {
	{"report", trace_report},
	{"snapshot", trace_snapshot},
	{"hist", trace_hist},
	{"mem", trace_mem},
	{"listen", trace_listen},
	{"agent", trace_agent},
	{"setup-guest", trace_setup_guest},
	{"split", trace_split},
	{"restore", trace_restore},
	{"stack", trace_stack},
	{"check-events", trace_check_events},
	{"record", trace_record},
	{"start", trace_start},
	{"set", trace_set},
	{"extract", trace_extract},
	{"stop", trace_stop},
	{"stream", trace_stream},
	{"profile", trace_profile},
	{"restart", trace_restart},
	{"clear", trace_clear},
	{"reset", trace_reset},
	{"stat", trace_stat},
	{"options", trace_option},
	{"show", trace_show},
	{"list", trace_list},
	{"help", trace_usage},
	{"dump", trace_dump},
	{"attach", trace_attach},
	{"convert", trace_convert},
	{"-h", trace_usage},
};

int main (int argc, char **argv)
{
	int i;

	errno = 0;

	if (argc < 2)
		trace_usage(argc, argv);

	for (i = 0; i < ARRAY_SIZE(commands); ++i) {
		if (strcmp(argv[1], commands[i].name) == 0 ){
			commands[i].run(argc, argv);
			goto out;
		}
	}

	/* No valid command found, show help */
	trace_usage(argc, argv);
out:
	exit(0);
}
